python/spacewalk/satellite_tools/appstreams.py

```
@@ -0,0 +1,297 @@
+gi.require_version("Modulemd", "2.0")
+from gi.repository import Modulemd
+
+# pylint: enable=wrong-import-position
+from spacewalk.satellite_tools.syncLib import log
+from spacewalk.server import rhnSQL
+
+
+class Nevra:
+    """
+    Represents a package NEVRA string.
+    """
+
+    def __init__(self, name: str, epoch: str, version: str, release: str, arch: str):
+        self.name = name
+        self.epoch = epoch
+        self.version = version
+        self.release = release
+        self.arch = arch
+
+    def __repr__(self):
+        epoch_str = f"{self.epoch}:" if self.epoch else ""
+        return f"{self.name}-{epoch_str}{self.version}-{self.release}.{self.arch}"
+
+
+class Nsvca:
+    """
+    Represents a NSVCA (name, stream, version, context, architecture) string.
+    """
+
+    def __init__(self, module):
+        """
+        Initializes an Nsvca object.
+
+        Parameters:
+        - module (Modulemd.ModuleStreamV2): Module stream object.
+        """
+        self.name = module.get_module_name()
+        self.stream = module.get_stream_name()
+        self.version = module.get_version()
+        self.context = module.get_context()
+        self.arch = module.get_arch()
+
+    def __repr__(self):
+        return f"{self.name}:{self.stream}:{self.version}:{self.context}:{self.arch}"
+
+
+class ModuleMdImporter:
+    """
+    Imports a channel's AppStream modules from its Modulemd file to the database.
+
+    Attributes:
+    - channel_id (int): ID of the channel.
+    - modulemd_file (str): Path to the channel's Modulemd file.
+    """
+
+    def __init__(self, channel_id: int, modulemd_file: str):
+        """
+        Initializes a ModuleMdImporter object.
+
+        Parameters:
+        - channel_id (str): ID of the channel.
+        - modulemd_file (str): Path to the Modulemd file.
+        """
+        self.channel_id = channel_id
+        self.modulemd_file = modulemd_file
+
+    def validate(self):
```

[nitpick] I would do the validation as part of the parsing. The current approach reads the same file twice.

---

The reason for that is that we want to validate the file at the beginning before starting to import packages and fail early. But the importing of modules must happen after the packages are linked because we need the package IDs at that point.

See the discussion: https://github.com/uyuni-project/uyuni-rfc/pull/87#discussion_r1540930114

---

Parsing the file early is fine for me, I just wonder why we parse it, throw it away and then parse it again at a later point.

---

Right, I get your point. I'll change it to use the already parsed index when importing at the end.